src/cpu/core_dyn_x86.cpp

```diff
@@ -275,21 +287,55 @@ static void copy_normal_fpu_to_dh()
 	FPU_GetPRegsTo(dyn_dh_fpu.state.st_reg);
 }
 
+static Bits copy_fpus_and_run_normal_core() noexcept
+{
+
+	// If host FPU has been used by the recompiled code
+	// block, then we need to save the host FPU state in
+	// dyn_dh_fpu.state and free up the host FPU for the
+	// normal core to use.
+	maybe_copy_host_fpu_to_dh();
+
+	// Copy the saved state to the normal core's emulated
+	// FPU so that it can resume from where the dynamic core
+	// left. We can't skip this step even when
+	// `dyn_dh_fpu.state_used` is `false`, because we do not
+	// know if the emulated FPU's state is consistent with
+	// the dynamic core, if not (very likely), then we will
+	// overwrite the dynamic FPU state later when we copy it
+	// back.
+	copy_dh_fpu_to_normal();
+
+	auto ret = CPU_Core_Normal_Run();
+
+	// On leaving the normal core, copy emulated FPU's state
+	// to the dynamic core's saved FPU state. There is no
+	// way to tell if the emulated FPU was used in this run
+	// so we can't skip this.
+	copy_normal_fpu_to_dh();
+
+	return ret;
```

Comment:
Nice explanatory comments 💯

Reply:
tried to summaries it more concisely in the latest push (and hopefully the code is more self-documenting too).